control_panel_api/filters.py

```diff
@@ -66,3 +67,27 @@ def filter_queryset(self, request, queryset, view):
             return queryset
 
         return queryset.filter(pk=request.user.pk)
+
+
+class UserS3BucketFilter(DjangoFilterBackend):
+    """
+    Filter to get visible UserS3Bucket records.
+
+    - Superusers see everything
+    - Normal users see only records for buckets they have access.
+      Normal users can also see records for other users that have access
+      to same data (but permissions will enforce they can't change these
+      records unless they're admin)
+    """
+
+    def filter_queryset(self, request, queryset, view):
+        queryset = super().filter_queryset(request, queryset, view)
+
+        if is_superuser(request.user):
+            return queryset
+
+        accessible_buckets = S3Bucket.objects.filter(
```

Comment:
As our base queryset is `UserS3Bucket.objects.all()` I think `queryset.filter(user=request.user)` should suffice.  

Not tested, just guessing.

Reply:
That was exactly the query I wrote the when I started to play with it but...

This is a bit of a sneaky one. Users need to see not only `UserS3Bucket` where the user is them but ALSO all other `UserS3Bucket` records for buckets they have access to. That's why the query is like that.

Let me know if it makes sense.